core: stdcm: conflict detection integration

eckter opened the pull request:
Fix https://github.com/osrd-project/osrd/issues/4484



The reason for such a massive PR is that we needed to have a stable testable product to be confident in the implementation. There has been several points where we thought "this part can be merged into dev" and yet we still had to make deep changes later on. 


Change summary:

1. Change the way we navigate the infra in STDCM
    1. Instead of going from block to block manually, we have an opaque object (`InfraExplorer`) that represent the different possible paths.
    1. Internally, it keeps track of routes and blocks from the start of the path, and can extend beyond the current point being evaluated by the pathfinding. 
    2. Another object built upon the first one can also keep track of past simulations / envelopes in a given path
1. Change the way conflicts are detected in STDCM
    1. Instead of manually identifying the available blocks, we call `SpacingRequirementAutomaton` and compare the results using `conflicts.kt` tools
    2. To do this, we needed a few tweaks to `conflicts.kt` to get the data we need. Because resource uses are supposed to be opaque, we needed to add a new method any time we needed to get something like minimum delay to add.
1.  Tests have been adapted
     1. The old classes used to identify available blocks have been kept and moved to the test modules. They are used to test STDCM while abstracting away the actual signaling. It's extremely useful to build the test scenarios in straightforward way.
     2. The conflict detection integration has mostly been tested at the `BlockAvailabilityInterface` level: we don't check how the conflicts are avoided (that's done in the tests mentioned above), but we check that the conflicting times are properly identified. 
1. Optimizations (not exhaustive because there are smaller ones):
    1. Cache lots of objects and results
    2. Avoid unnecessary object instantiations and computations
    3. Spacing resource requirements: use a binary search to find the last required zone
    4. Avoid instantiating edges that would be considered "already visited"



**About the git history**: I did my best to clean it up, but there's only so much that can be done. Commit-by-commit review is not recommended **for STDCM files**. There has been whole classes that have been split and re-merged and tweaked, and too much stuff happened in the meantime for proper fixups/squashs. Changes that are outside of STDCM scope **have clear atomic commits** (generally bugfixes), but new files and methods should be reviewed in one go. I've kept most of the history because it may still help a little with the review, but I'll do a more agressive squash towards the end of the process. The plan is to squash everything up to the point where we transition to bugfixes / optimizations only. Once reviews start coming, we'll stick to fixup commits and avoid rebases.

This has been tested with extensive fuzzing on both small_infra and France, including a new assertion that the new train isn't conflicting with the timetable. The bugs that aren't already present on dev have been fixed. There's also a fairly large amount of unit tests in core (IIRC more than half of the line diff is located in test files). I also tested manually on small_infra and on France, but not in depth. **We need to test once again because of the optimization changes**, but we can still start the reviews

Additional notes:

1. Some implementations have poor performances. We did our best to keep the PR straightforward, optimization will likely be planned for the next PI
    1. edit: we do have performances issues, but the poor complexity in the new classes have nothing to do with it
5. Similarly, for the most part we only consider the spacing conflicts. Routing conflicts will be added later on (it *should* be mostly painless)
6. Deadline: end of this week :'( (16/03)
7. Once approved, I'll update the documentation on the website

eckter committed:
core: stdcm: implement and use new interface for infra exploration InfraExplorer

tests: add a new infra with overlapping routes

core: stdcm: add tests for infra explorer

core: implement PathPropertiesView

core: implement new path exploration class InfraExplorer

core: stdcm: implement and use new interface for infra exploration InfraExplorer

eckter committed:
core: implement and use InfraExplorerWithEnvelope

Erashin commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/utils/PathPropertiesView.kt:
```diff
@@ -0,0 +1,93 @@
+    val base: PathProperties,
+    val startOffset: Offset<Path>,
+    val endOffset: Offset<Path>
+) : PathProperties {
```
```suggestion
): PathProperties {
```

Erashin commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/utils/InfraUtils.kt:
```diff
@@ -20,3 +16,111 @@ fun TrackNetworkInfra.getNextTrackSections(trackSection: DirTrackSectionId): Dir
+    blockInfra: BlockInfra,
+    routeId: RouteId
+): Boolean {
+    val routeIds = MutableStaticIdxArrayList<Route>()
```
You could combine them into `mutableStaticIdxArrayListOf(routeId)`.

Erashin reviewed (changes requested):
Lgtm up to this point :)

eckter committed:
core: stdcm: fix block availability legacy adapter

eckter committed:
core: stdcm: implement new version of BlockAvailabilityInterface

eckter committed:
core: stdcm: infra_explorer: adjust start offset when using envelope

eckter committed:
core: stdcm: handle "not enough lookahead" responses

eckter committed:
core: stdcm: add end of path to infra explorer

Erashin committed:
core: stdcm: implement occupancy requirement tests

eckter committed:
core: stdcm: keep one spacing automaton per explorer

eckter committed:
core: stdcm: always extend infra explorers lookahead till end of route

The reasoning is to reduce the number of `NotEnoughLookahead`
exceptions, there's no need to restrict the lookahead when we
know the direction

Erashin committed:
core: make SpacingResourceGenerator cloneable

eckter committed:
core: stdcm: adjust conflict detection module to fit stdcm and use it in BlockAvailability

Co-authored-by: Eloi Charpentier <eloi.charpentier.42@gmail.com>

eckter committed:
core: stdcm: write tests for block availability

eckter committed:
core: stdcm: use new conflict detection

eckter committed:
core: stdcm: explorer: change path property cache to lazy evaluation

We were missing the entries for the blocks on the first route,
this change is the easiest way to fix

eckter committed:
core: only add unprotected zone requirement when actually reached

This caused a bug in an edge case, when the train
starts at a place where a signal is seen but its
zone not reached

eckter committed:
core: change "routesOnBlock" to only include routes with the block

eckter committed:
core: infra-explorer: include blocks before the start one

We used to not consider the blocks on the routes before the
start one, but as a result the conversions to/from
travelled path offsets were wrong

eckter committed:
core: stdcm: add envelope even when stopping in prev block

eckter committed:
core: stdcm: remove unavailable space builder

eckter committed:
core: stdcm: move dummy block occupancy to test modules

eckter committed:
core: stdcm: add doc to explorer

eckter committed:
core: signaling: change return type from idxMap to map

The map returned from the signaling simulator is
sparse, which IdxMap can *not* handle in an efficient way

eckter committed:
core: stdcm: always extend lookahead until 4 blocks

The value is arbitrary but seems to fit for BAL,
I tried to increase it one by one until no exception
is thrown.

eckter committed:
core: stdcm: keep spacing automaton state over the edges

eckter committed:
core: stdcm: reduce the number of object allocations

eckter committed:
core: spacing-resources: split main loop into distinct methods

core: spacing-generator: fix accidental n^2 complexity on repeated calls

core: signaling simulator: avoid redundant block list copy

The previous version inputed the whole path to each call
of the signaling simulation, resulting on a very poor
scaling on long paths in stdcm

eckter commented on core/src/main/java/fr/sncf/osrd/conflicts/SpacingResourceGenerator.kt:
```diff
@@ -201,35 +179,34 @@ class SpacingRequirementAutomaton(
+            val blocks = mutableStaticIdxArrayListOf(incrementalPath.getBlock(firstBlockIndex))
+            var nSimulatedZones =
+                blockInfra.getBlockPath(blocks[0]).size // number of zones in the blocks
+            val firstZone = incrementalPath.getBlockStartZone(firstBlockIndex)
+            var lastZone = firstZone + 1 // excluded
```
rename to more precise names, include it

multun replied:
lastZone is a weird name for something we're iterating on. It also seems to be the exact same as nextProbedZoneForSignal

multun commented on core/src/main/java/fr/sncf/osrd/conflicts/SpacingResourceGenerator.kt:
```diff
@@ -110,12 +105,6 @@ class SpacingRequirementAutomaton(
-    private fun getSignalProtectedZone(signal: PathSignal): Int {
-        // the signal protects all zone inside the block
-        // if a signal is at a block boundary,
-        return incrementalPath.getBlockEndZone(signal.minBlockPathIndex)
```
I believe this method behaves properly. There are two main cases:
 - either the signal is a block delimiter, like a BAL3 signal. in this case, minBlockPathIndex returns the index of block which ends at this signal. Then getBlockEndZone returns the index of the zone after the block.
 - if the signal does not delimit a block, like a BAPR distant signal, minBlockPathSignal is the current block, and getBlockEndZone is the index of the next zone after the block

I believe the documentation of getSignalProtectedZone should be improved, and the documentation of getBlockEndZone should make clear it returns an excluded bound

multun commented on core/src/main/java/fr/sncf/osrd/conflicts/SpacingResourceGenerator.kt:
```diff
@@ -239,17 +216,16 @@ class SpacingRequirementAutomaton(
                         0,
```
I believe a lot of the setup you're doing could be avoided by using the start block parameter of simulator.evaluate. If it turns out to not be useful, we could also remove the parameter entirely and make the code simpler

multun commented on core/src/main/java/fr/sncf/osrd/conflicts/SpacingResourceGenerator.kt:
```diff
@@ -201,35 +179,34 @@ class SpacingRequirementAutomaton(
-                if (nextProbedZoneForSignal == zoneEndIndex) {
-                    if (incrementalPath.pathComplete) break@zoneProbingLoop
-                    return NotEnoughPath
+                // setup the block / zone arrays
```
I'd love a comment as follows:

In the process of figuring out whether a zone's occupancy would disturb the train, we need to simulate more and more blocks along the way.

eckter committed:
core: envelope-concat: directly use last envelope for total time

eckter committed:
core: stdcm: ignore lookahead when caching sim results

eckter committed:
core: envelope: use binary search to find the right envelope

eckter committed:
core: infra-explorer: cache concatenated envelopes

eckter committed:
core: spacing-resource: use a binary search

eckter committed:
core: stdcm: keep located envelopes in infra explorer

eckter committed:
core: stdcm: fix bug on "seen" edges with stops

Repeated edges at different locations of the same
block were considered identical

eckter committed:
core: stdcm: filter out seen edges before instantiating them

eckter committed:
core: stdcm: block-availability: add bound to binary search

We could sometimes enter infinite loops on stop times

Erashin commented on core/src/test/kotlin/fr/sncf/osrd/stdcm/FullSTDCMTests.kt:
```diff
@@ -135,21 +130,15 @@ class FullSTDCMTests {
-            )
-        )
+        val requirements = makeRequirementsFromPath(infra, start, end, 0.0).toMutableList()
+        val minDelay = getMaxOccupancyLength(requirements)
```
`getMaxOccupancyDuration`?

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/utils/InfraUtils.kt:
```diff
@@ -20,3 +17,149 @@ fun TrackNetworkInfra.getNextTrackSections(
+
+    // We need to evaluate the longest route first, in case one route covers a subset of another
+    val sortedRoutes = routes.sortedBy { r -> -infra.getChunksOnRoute(r).size }
+    for (routeId in sortedRoutes) if (
+        routeMatchPath(infra, blockInfra, chunks, startIndex, routeMustIncludeStart, routeId)
+    )
```
that's some very cursed formatting, can you add brackets to help the formatter unfuck it?

multun commented on core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/utils/InfraUtils.kt:
```diff
@@ -20,3 +17,149 @@ fun TrackNetworkInfra.getNextTrackSections(
+    routeId: RouteId
+): Boolean {
+    var mutChunkIndex = chunkIndex
+    if (!routeHasBlockPath(infra, blockInfra, routeId))
+        return false // Filter out routes that don't have block, they would cause issues later on
```
I know that's not quite related, but I don't know if I'm comfortable having this check here: what does "later on" mean? sure, in some contexts, a route may need to have both blocks and match a given path. But then this is best checked in this specific context.

eckter replied:
"later on" would mean "any time we call `recoverBlocks` and assert that there is a result", which is quite often (and not actually STDCM-related, this has just been moved). Most of the code has been written with the assumption that every route has a block path. 

In my opinion these routes should be discarded as part of the loading process and not part of the infra, but I don't know how that would be implemented (considering the blocks are loaded after the routes)

multun replied:
> Most of the code has been written with the assumption that every route has a block path

I think that's the core issue here. Because not all blocks types can be used by all trains, I don't see how that's relevant: even if a route is entirely covered by blocks, they may not be usable

eckter replied:
I'd say fixing this is out of scope here, it's not directly related and I think it would be a fairly large change. I can open an issue.

eckter replied:
https://github.com/osrd-project/osrd/issues/6884

multun commented on core/src/main/java/fr/sncf/osrd/conflicts/Conflicts.kt:
```diff
@@ -37,9 +41,29 @@ fun detectConflicts(trainRequirements: List<TrainRequirements>): List<Conflict>
 
     fun checkRoutingRequirement(req: RoutingRequirement): List<Conflict>
+
+    fun minDelayWithoutConflicts(
+        spacingRequirements: List<SpacingRequirement>,
+        routingRequirements: List<RoutingRequirement>
+    ): Double
+
+    fun maxDelayWithoutConflicts(
+        spacingRequirements: List<SpacingRequirement>,
+        routingRequirements: List<RoutingRequirement>
+    ): Double
+
+    fun timeOfNextConflict(
+        spacingRequirements: List<SpacingRequirement>,
+        routingRequirements: List<RoutingRequirement>
+    ): Double
```
wouldn't it be faster to return all properties at once? I can think of having a analyzeConflicts() method which would return an object with all of these properties

multun commented on core/src/main/java/fr/sncf/osrd/conflicts/SpacingResourceGenerator.kt:
```diff
@@ -172,84 +170,127 @@ class SpacingRequirementAutomaton(
+    // Returns the index of the first zone that isn't required for the given signal,
+    // or null if we need more path to determine it
+    private fun findFirstNonRequiredZoneIndex(pathSignal: PathSignal): Int? {
+        // We are looking for the index `i` where `isZoneIndexRequiredForSignal` returns
+        // true at `i-1` and false at `i`. We could just iterate starting at 0, but
+        // because `i` is not that small (20 on average) and the signaling
+        // simulation calls are expensive, we prefer an approach similar to
+        // a binary-search
+
+        // The values here are determined empirically on imported infrastructures,
+        // the solutions mostly follow a gaussian distribution centered
+        // on start+20 that rarely exceeds start+40.
+        // We run a binary search on that range, and iterate one by one when the solution is above.
```
Does this really yield any performance benefit at all? What if the bound exceeds 40?

eckter replied:
Roughly -30% execution time, for an arbitrary request on France repeated 15 times on each version. We go from 20 to 5 simulator calls per signal on average.

I do have some plans to improve it further, this loop is one of the main execution time bottlenecks even with this change. But this PR is large enough as it is.

If the bound exceeds 40 it moves forward zone by zone (as it did before).

multun commented on core/src/main/java/fr/sncf/osrd/standalone_sim/result/ResultTrain.java:
```diff
@@ -89,7 +89,6 @@ public SpacingRequirement(String zone, double beginTime, double endTime, boolean
             assert !Double.isNaN(beginTime);
             assert !Double.isNaN(endTime);
             assert Double.isFinite(beginTime);
-            assert Double.isFinite(endTime);
```
is it because of partial requirements?

eckter replied:
It's for tests, we often want to input spacing requirements that have no end time

eckter committed:
core: stdcm: use an append-only linked list for path data

We used to copy every data about the past elements of the
path when we explore diverging paths, we now keep
linked list that don't duplicate data

eckter commented on core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/DelayManager.kt:
```diff
@@ -131,29 +116,18 @@ internal constructor(
         startTime: Double,
-        stopDurationAtEnd: Double?
     ): Availability {
-        val speedRatio = graph.getStandardAllowanceSpeedRatio(envelope)
```
note to self, this should not have been removed

eckter committed:
core: stdcm: rename/document methods handling standard allowance

eckter committed:
core: spacing-resources: break off when path offset = sight offset

When both are equal we enter into the "epsilon error"
ranges of float operations in `callbacks.arrivalTimeInRange`

eckter reviewed (commented):
I can't approve my own PR, but for what it's worth I've re-read everything and it looks good to me

Erashin reviewed (approved):
LGTM at this point. Yay :)

multun commented on core/src/main/java/fr/sncf/osrd/conflicts/IncrementalPath.kt:
```diff
@@ -36,25 +35,13 @@ class PathFragment(
-    val beginZonePathIndex: Int
-    val beginBlockIndex: Int
-    val beginRouteIndex: Int
+    fun clone(): IncrementalPath
```
I think the current implementation is _alright_, but I can also see a cheaper structural sharing opportunity: instead of having separate linked lists for each field, IncrementalPath itself can be a linked list, where property accesses to fields before the current node's range are delegated to the next node

eckter replied:
The current implementation sounded less "intrusive" that refactoring the incremental path as a whole, and we needed a linked lists for other elements of `InfraExplorer` anyway.

I do think making IncrementalPath itself a linked list would be better, but that can be for later

multun reviewed (approved):
I left some minor comments, which can be addressed later
